src/libraries/Microsoft.Extensions.Logging/gen/TemplateHelper.cs

```diff
@@ -0,0 +1,346 @@
+                        {
+                            // dangling }, fail
+                            throw new ArgumentException($"Dangling }} in format string at position {pos}", nameof(format));
+                        }
+                    }
+                    else if (ch == '{')
+                    {
+                        if (pos < len && format[pos] == '{')
+                        {
+                            // double {, treat as escape sequence
+                            pos++;
+                        }
+                        else
+                        {
+                            // start of a format specification
+                            pos--;
+                            break;
+                        }
+                    }
+
+                    builder.Append(ch);
+                }
+
+                if (pos == len)
+                {
+                    var totalLit = builder.Length - segStart;
+                    while (totalLit > 0)
+                    {
+                        var num = totalLit;
+                        if (num > short.MaxValue)
+                        {
+                            num = short.MaxValue;
+                        }
+
+                        segments.Add(new Segment((short)num, -1, 0, string.Empty));
+                        totalLit -= num;
+                    }
+
+                    return builder.ToString();
+                }
+
+                // extract the argument index
+                var argIndex = 0;
+                if (templates == null)
+                {
+                    // classic composite format string
+
+                    pos++;
+                    if (pos == len || (ch = format[pos]) < '0' || ch > '9')
+                    {
+                        // we need an argument index
+                        throw new ArgumentException($"Missing argument index in format string at position {pos}", nameof(format));
+                    }
+
+                    do
+                    {
+                        argIndex = (argIndex * 10) + (ch - '0');
+                        pos++;
+
+                        // make sure we get a suitable end to the argument index
+                        if (pos == len)
+                        {
+                            throw new ArgumentException($"Invalid argument index in format string at position {pos}", nameof(format));
+                        }
+
+                        ch = format[pos];
+                    }
+                    while (ch >= '0' && ch <= '9');
+                }
+                else
+                {
+                    // template-based format string
+
+                    pos++;
+                    if (pos == len)
+                    {
+                        // we need a template name
+                        throw new ArgumentException($"Missing template name in format string at position {pos}", nameof(format));
+                    }
```

Comment:
you can move this block before the condition 
if (templates == null) so you don't have to repeat the same code